test/e2e/lifecycle.go

```diff
@@ -0,0 +1,217 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"knative.dev/pkg/apis"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	"knative.dev/pkg/test"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/pkg/test/helpers"
+)
+
+// Setup creates the client objects needed in the e2e tests,
+// and does other setups, like creating namespaces, set the test case to run in parallel, etc.
+func Setup(t *testing.T, runInParallel bool) *Client {
+	// Create a new namespace to run this test case.
+	baseName := helpers.AppendRandomString(helpers.GetBaseFuncName(t.Name()))
+	namespace := helpers.MakeK8sNamePrefix(baseName)
+	t.Logf("namespace is : %q", namespace)
+	client, err := NewClient(
+		pkgTest.Flags.Kubeconfig,
+		pkgTest.Flags.Cluster,
+		namespace,
+		t)
+	if err != nil {
+		t.Fatalf("Couldn't initialize clients: %v", err)
+	}
+
+	client.CreateNamespaceIfNeeded(t)
+	client.DuplicateSecret(t, "google-cloud-key", "default")
+
+	// Disallow manually interrupting the tests.
+	// TODO(Fredy-Z): t.Skip() can only be called on its own goroutine.
+	//                Investigate if there is other way to gracefully terminte the tests in the middle.
+	c := make(chan os.Signal)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		fmt.Printf("Test %q running, please don't interrupt...\n", t.Name())
+	}()
+
+	// Run the test case in parallel if needed.
+	if runInParallel {
+		t.Parallel()
+	}
+
+	return client
+}
+
+// TearDown will delete created names using clients.
+func TearDown(client *Client) {
+	if err := DeleteNameSpace(client); err != nil {
+		client.T.Logf("Could not delete the namespace %q: %v", client.Namespace, err)
+	}
+}
+
+// DeleteNameSpace deletes the namespace that has the given name.
+func DeleteNameSpace(client *Client) error {
+	_, err := client.Kube.Kube.CoreV1().Namespaces().Get(client.Namespace, metav1.GetOptions{})
+	if err == nil || !errors.IsNotFound(err) {
+		return client.Kube.Kube.CoreV1().Namespaces().Delete(client.Namespace, nil)
+	}
+	return err
+}
+
+// Client holds instances of interfaces for making requests to Knative.
+type Client struct {
+	Kube    *test.KubeClient
+	Dynamic dynamic.Interface
+
+	Namespace string
+	T         *testing.T
+}
+
+// NewClient instantiates and returns several clientsets required for making request to the
```

Comment:
this comment seems a bit wrong? 'several clientsets' => just a single Client?

Reply:
copy pasta from eventing... sorry